include/mbgl/storage/file_source.hpp

```diff
@@ -29,22 +45,44 @@ class FileSource {
     // not be executed.
     virtual std::unique_ptr<AsyncRequest> request(const Resource&, Callback) = 0;
 
+    // Allows to forward response from one source to another.
+    // Optionally, callback can be provided to receive notification for forward
+    // operation.
+    virtual void forward(const Resource&, const Response&, ForwardCallback = {}) {}
+
     // When a file source supports consulting a local cache only, it must return true.
     // Cache-only requests are requests that aren't as urgent, but could be useful, e.g.
     // to cover part of the map while loading. The FileSource should only do cheap actions to
     // retrieve the data, e.g. load it from a cache, but not from the internet.
-    virtual bool supportsCacheOnlyRequests() const {
-        return false;
-    }
+    virtual bool supportsCacheOnlyRequests() const { return false; }
+
+    // Checks whether a resource could be requested from this file source.
+    virtual bool canRequest(const Resource&) const = 0;
+
+    /*
+     * Pause file request activity.
+     *
+     * If pause is called then no revalidation or network request activity
+     * will occur.
+     */
+    virtual void pause() {}
+
+    /*
+     * Resume file request activity.
+     *
+     * Calling resume will unpause the file source and process any tasks that
+     * expired while the file source was paused.
+     */
+    virtual void resume() {}
 
-    // Singleton for obtaining the shared platform-specific file source. A single instance of a file source is provided
-    // for each unique combination of a Mapbox API base URL, access token, cache path and platform context.
-    static std::shared_ptr<FileSource> getSharedFileSource(const ResourceOptions&);
+    /*
+     * Generic property setter / getter methods.
+     */
+    virtual void setProperty(const std::string&, const mapbox::base::Value&){};
```

Comment:
I suppose this is not typed to allow user implementations to extend this right? I like the idea but we could at least use some defines for our standard properties.

Reply:
@tmpsantos what would be a good place to keep those? constants.hpp?